### PR TITLE
Adding JS tests for sync up merge modes

### PR DESCRIPTION
**Pull request by bhariharan**

### CODE REVIEW

File: test/SFSmartSyncTestSuite.js [hunk 1]
```diff
@@ -2810,6 +2810,77 @@ SmartSyncTestSuite.prototype.testSyncUpLocallyUpdated = function() {
...
+        })
+        .then(function() {
+            console.log("## Updating local records");
+            updatedRecords = [];
+            _.each(_.keys(idToName), function(id) {
+                updatedRecords.push({Id:id, Name:idToName[id]+"Updated", __locally_updated__:true});
+            });
+            return cache.saveAll(updatedRecords);
+        })
+        .then(function() {
```
Comment (bhariharan) [inline test/SFSmartSyncTestSuite.js]: Same as the other ```syncUp``` test, except this addition here, to edit the server record before attempting to sync up again.

File: test/SFSmartSyncTestSuite.js [hunk 2]
```diff
@@ -2871,6 +2942,77 @@ SmartSyncTestSuite.prototype.testSyncUpLocallyDeleted = function() {
...
+            console.log("## Deleted local records");
+            deletedRecords = [];
+            _.each(_.keys(idToName), function(id) {
+                deletedRecords.push({Id:id, __locally_deleted__:true});
+            });
+            return cache.saveAll(deletedRecords);
+        })
+        .then(function() {
```
Comment (bhariharan) [inline test/SFSmartSyncTestSuite.js]: Same as the other ```syncUp``` test, except this addition here, to edit the server record before attempting to sync up again.

**Comment by khawkins** [comment]
Seems fine to me.

File: test/SFSmartSyncTestSuite.js [hunk 3]
```diff
@@ -2756,7 +2756,7 @@ SmartSyncTestSuite.prototype.testSyncUpLocallyUpdated = function() {
     var self = this;
     var idToName = {};
     var updatedRecords;
-    var options = {fieldlist: ["Name"]};
+    var options = {fieldlist: ["Name"], mergeMode: Force.MERGE_MODE_DOWNLOAD.OVERWRITE};
```
Comment (wmathurin) [inline test/SFSmartSyncTestSuite.js]: Maybe the enum Force.MERGE_MODE_DOWNLOAD needs to be renamed since we are using it also for upload?

Comment (bhariharan) [inline test/SFSmartSyncTestSuite.js]: Sure - I'll make that change.

**Comment by bhariharan** [comment]
@wmathurin I replaced ```MERGE_MODE_DOWNLOAD``` with ```MERGE_MODE``` in [this](https://github.com/bhariharan/SalesforceMobileSDK-Shared/commit/c6d7f776907d33a8f896ddc45f29c337e4817e90) commit.

**Comment by wmathurin** [comment]
There already was a Force.MERGE_MODE defined at line 929 with different value - it is used for conflict behavior on save.

That's why I introduced a separate enum for the download merge mode. The download merge mode values are in line with what java/objective-c expect.

MERGE_MODE_DOWNLOAD was actually used by StoreCache in smartsync.js. Will the upload one be used in smartsync.js or only in the mock plugin to mock a sync up with leave-if-changed?

Here are my suggestions, let's define a MERGE_MODE enum as part of the SmartSyncPlugin (exported on the plugin) - let's keep MERGE_MODE_DOWNLOAD in smartsync.js with a note saying that the string values should line up with the plugin. For sync up merge mode, you should only change MockSmartSyncPlugin.

**Comment by bhariharan** [comment]
@wmathurin I'm going to back out the changes in my last commit, and make the changes you suggested for merge modes in a separate pull request if that's alright.

**Pull request merged**